File: python/ch1/p2.py
"""
 CREATED BY: Benjamin Hamlin
 Chapter 1 Question 2
 "Check Permutation"
 PROMPT:
 Given two stings, write a method to decide if one is a permutation of the other
 DATE: 11/14/2019
"""
import unittest

# O(N) where N is the length of str1
# O(1) when strings are not the same length


def p2(str1, str2):
    if len(str1) == len(str2):
        for letter in str1:
            if str1.count(letter) != str2.count(letter):
                return False
        return True
    else:
        return False


class TestP2(unittest.TestCase):
    def test_p2(self):
        self.assertTrue(p2("cat", "tac"))
        self.assertTrue(p2("own", "won"))
        self.assertFalse(p2("chicken", "butt"))


if __name__ == '__main__':
    temp = input("Enter a string:")
    temp2 = input("Enter a string to see if it is a permutation of the other:")
    print(p2(temp, temp2))

File: c/ch1/p2.c
/*
CREATED BY: Benjamin Hamlin
Chapter 1 Question 2
"Check Permutation"
PROMPT:
Given two stings, write a method to decide if one is a permutation of the other
DATE: 11/23/2019
*/

#include "header.h"

// O(n) where n is the length of the strings
int p2(char* s1, char* s2){
   int alpha[93] = {0};
   if(strlen(s1) == strlen(s2)){
      for(int i = 0; s1[i] != '\0'; ++i){
         if(s1[i] != ' '){
            alpha[s1[i]-'!']++;
         }
         if(s2[i] != ' '){
            alpha[s2[i]-'!']++;
         }
      }
      // O(1)
      for(int i = 0; i < 93; ++i){
         if(alpha[i]%2 != 0){
            return 0;
         }
      }
      return 1;
   }
   return 0;
}

int main(int argc, char* argv[]){
   if(argc > 1){
      if(strcmp("-d", argv[1]) == 0 ){
         if(p2("care", "race") == 1){
            printf("ok");
         }
         else{
            printf("Failed");
         }
      }
   }
}
